src/exclusion/input_filter.cpp

```diff
@@ -11,44 +11,28 @@ namespace ddwaf::exclusion {
 
 using excluded_set = input_filter::excluded_set;
 
-input_filter::input_filter(std::string id, std::vector<condition::ptr> conditions,
-    std::set<rule *> rule_targets, std::shared_ptr<object_filter> filter)
-    : id_(std::move(id)), conditions_(std::move(conditions)),
-      rule_targets_(std::move(rule_targets)), filter_(std::move(filter))
-{}
+input_filter::input_filter(std::string id, expression::ptr expr, std::set<rule *> rule_targets,
+    std::shared_ptr<object_filter> filter)
+    : id_(std::move(id)), expr_(std::move(expr)), rule_targets_(std::move(rule_targets)),
+      filter_(std::move(filter))
+{
+    if (!expr_) {
+        throw std::invalid_argument("rule filter constructed with null expression");
+    }
+}
 
 std::optional<excluded_set> input_filter::match(
     const object_store &store, cache_type &cache, ddwaf::timer &deadline) const
 {
-    if (!cache.result) {
-        std::vector<condition::ptr>::const_iterator cond_iter;
-        bool run_on_new;
-        if (cache.last_cond.has_value()) {
-            cond_iter = *cache.last_cond;
-            run_on_new = true;
-        } else {
-            cond_iter = conditions_.cbegin();
-            run_on_new = false;
+    // An event was already produced, so we skip the rule
+    // Note that conditions in a filter are optional
+    if (!expr_->empty() && !expression::get_result(cache.expr_cache)) {
+        if (!expr_->eval(cache.expr_cache, store, {}, {}, deadline)) {
```

Comment:
You can merge `ifs` here

Reply:
I'll make the change in https://github.com/DataDog/libddwaf/pull/194 since it's derived from this one.